object_database/web/ActiveWebService.py

```diff
@@ -27,34 +27,130 @@
 import json
 import gevent.socket
 
-from gevent import pywsgi
-from geventwebsocket.handler import WebSocketHandler
-
+from object_database.view import revisionConflictRetry
+from object_database.util import genToken, checkLogLevelValidity
+from object_database import ServiceBase, service_schema, Schema, Indexed, Index, DatabaseObject
+from object_database.web.AuthPlugin import AuthPluginBase, PermissiveAuthPlugin, LdapAuthPlugin
 from object_database.web.cells import *
+from typed_python import TupleOf, Float64
 
+from gevent import pywsgi, sleep
 from gevent.greenlet import Greenlet
-from gevent import sleep
+from geventwebsocket.handler import WebSocketHandler
 
-from flask import Flask, send_from_directory, redirect, url_for, request
+from flask import Flask, send_from_directory, redirect, url_for, request, render_template, flash
+from flask_wtf import FlaskForm
 from flask_sockets import Sockets
 from flask_cors import CORS
+from flask_login import LoginManager, current_user, login_user, logout_user, login_required
+
+from wtforms import StringField, PasswordField, BooleanField, SubmitField
+from wtforms.validators import DataRequired
 
-from object_database import ServiceBase, service_schema, Schema, Indexed, Index, DatabaseObject
 
 active_webservice_schema = Schema("core.active_webservice")
 
+
 @active_webservice_schema.define
 class Configuration:
     service = Indexed(service_schema.Service)
 
     port = int
     hostname = str
 
+    log_level = int
+
+    # auth plugin args
+    auth_type = str
+    authorized_groups = TupleOf(str)
+    auth_hostname = str
+    ldap_base_dn = str
+    ldap_ntlm_domain = str
+
+    # HTML template rendering args
+    company_name = str
+
+
+USER_LOGIN_DURATION = 24 * 60 * 60  # 24 hours
+
+@active_webservice_schema.define
+class User:
+    username = Indexed(str)
+    login_expiration = Float64()
+
+    def login(self):
+        self.login_expiration = time.time() + USER_LOGIN_DURATION
+
+    def logout(self):
+        self.login_expiration = 0.0
+
+
+class UserWrapper:
+    #  must be called with an open view of the user argument
+    @staticmethod
+    def makeFromUser(user):
+        if user is None:
+            return None
+        else:
+            return UserWrapper(user.username, user.login_expiration)
+
+    def __init__(self, username, login_expiration):
+        self.username = username
+        self.login_expiration = login_expiration
+
+    @property
+    def is_authenticated(self):
+        return True if time.time() < self.login_expiration else False
```

Comment:
you might want to have the 'db' here instead of the login_expiration, and check the value directly from a view rather than stashing it in the object. Otherwise, this object will become stale. If flask is not stashing this anywhere (that is, if it's recreating this object every time we use it) then it won't be a problem. But as it stands, anyone who has this object could get an out-of-date view of the ttl if they just hold onto the object.

Reply:
I though about that but the model is that the `flask-login` plugin assumes that the User objects live in some sort of DB (typically a relational DB) and that's why it requires you to register a callback with it for `load_user` that it calls each time it needs to access a User field, for example `is_authenticated` (I checked that by adding print statements to the `load_user` call back and the `is_authenticated` property). So, I thought that this dumb UserWrapper design is better because it is simpler.